### PR TITLE
Fix case sensitivity rules for wildcard queries on text fields

### LINKED ISSUE #71403: Wildcard queries on text fields don't obey rules for case sensitivity
A `wildcard` query for `*Butterflies*` should not match a `text` field with the value `I like butterflies`.

This used to be the behaviour before 7.8 but then [this fix](https://github.com/elastic/elasticsearch/pull/53127) to `wildcard` queries on `keyword` fields inadvertently changed the behaviour for `text` fields. Now it always normalizes the search string due to changes in the common base class (StringFieldType).

The solution should be to only normalize the search string when the new `case_insensitive` query flag is set to true.

### PR DESCRIPTION
**Pull request by cbuescher**
Wildcard queries on text fields should not apply the fields analyzer to the
search query. However, we accidentally enabled this in #53127 by moving the
query normalization to the StringFieldType super type. This change fixes this by
separating the notion of normalization and case insensitivity (as implemented in
the `case_insensitive` flag). This is done because we still need to maintain
normalization of the query sting when the wildcard query method on the field type is
requested from the `query_string` query parser. Wildcard queries on keyword
fields should also continue to apply the fields normalizer, regardless of
whether the `case_insensitive` is set, because normalization could involve
something else than lowercasing (e.g. substituting umlauts like in the
GermanNormalizationFilter).

Closes #71403

### CODE REVIEW

File: server/src/internalClusterTest/java/org/elasticsearch/search/query/SearchQueryIT.java [hunk 1]
```diff
@@ -1806,7 +1806,7 @@ public void testWildcardQueryNormalizationOnTextField() {
          {
              WildcardQueryBuilder wildCardQuery = wildcardQuery("field1", "Bb*");
              SearchResponse searchResponse = client().prepareSearch().setQuery(wildCardQuery).get();
-             assertHitCount(searchResponse, 1L);
+             assertHitCount(searchResponse, 0L);
 
              wildCardQuery = wildcardQuery("field1", "bb*");
              searchResponse = client().prepareSearch().setQuery(wildCardQuery).get();
```
Comment (markharwood) [inline server/src/internalClusterTest/java/org/elasticsearch/search/query/SearchQueryIT.java:1812]: Maybe add a test where the search string is mixed case but set `wildCardQuery.caseInsensitive(true)`

File: server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java [hunk 2]
```diff
@@ -322,6 +324,19 @@ protected BytesRef indexedValueForSearch(Object value) {
...
+         * Wildcard queries on keyword fields use the normalizer of the underlying field, regardless of their case sensitivity option
+         */
+        @Override
+        public Query wildcardQuery(
+            String value,
+            MultiTermQuery.RewriteMethod method,
+            boolean caseInsensitive,
+            SearchExecutionContext context
+        ) {
+            return super.wildcardQuery(value, method, caseInsensitive, true, context);
```
Comment (markharwood) [inline server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:337]: Maybe calling `normalizedWildcardQuery` here would make the intent more obvious plus help any tracing back of where we make use of normalized wildcard queries

**Review by markharwood — COMMENTED** [review]
Left a couple of comments and it also looks like WildcardFieldMapper is missing a `normalizedWildcardQuery` override but otherwise LGTM

File: server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java (hunk 2 above)
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:337]: Good idea, but I left the `caseInsensitive` parameter out of the `normalizedWildcardQuery` signature on purpose, so I need to call the protected method that takes both arguments here. My thinking was that we use `normalizedWildcardQuery` only from QueryStringQueryParser where we don't have the `caseInsensitive` option. Maybe you see a different solution?

Comment (markharwood) [reply server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:337]: Ah ok. That makes sense.

**Pull request merged**